lib/puppet/settings.rb

```diff
@@ -146,8 +147,8 @@ def initialize
     @configuration_file = nil
 
     # And keep a per-environment cache
-    @cache = Hash.new { |hash, key| hash[key] = {} }
-    @values = Hash.new { |hash, key| hash[key] = {} }
+    @cache = Concurrent::Hash.new { |hash, key| hash[key] = Concurrent::Hash.new }
+    @values = Concurrent::Hash.new { |hash, key| hash[key] = Concurrent::Hash.new }
```

Comment:
I think this may not work as expected when environment timeout is enabled, because if a thread marks an environment as expired, then only that thread's cache will be cleared. Requests handled by a different thread will still see the old realpath-ed `modulepath`, I think?